lib/events.ml

```diff
@@ -0,0 +1,380 @@
+open Common
+open Lwt.Infix
+
+module Smap = Map.Make(String)
+
+module Message = struct
+  type recipients =
+    | Stream of int * string
+    | Private of int list
+
+  type flag =
+    | Read 
+    | Starred
+    | Collapsed
+    | Mentioned
+    | Wildcard_mentioned
+    | Has_alert_word
+    | Historical
+
+  let flag =
+    Json_encoding.(
+      string_enum [
+        "read", Read;
+        "starred", Starred;
+        "collapsed", Collapsed;
+        "mentioned", Mentioned;
+        "wildcard_mentioned", Wildcard_mentioned;
+        "has_alert_word", Has_alert_word;
+        "historical", Historical;
+      ]
+    )
+
+  type t = { 
+    id: int ;
+    sender_id: int ;
+    sender_email: string ;
+    sender_full_name: string;
+    recipients: recipients ;
+    content : string ;
+    flags : flag list ;
+  }
+
+  let content { content; _ } = content
+
+  let sender_mention ?(silent = false) { sender_full_name; sender_id; _ } =
+    Format.asprintf "@%s**%s|%d**" (if silent then "_" else "") sender_full_name sender_id
+
+  let sender_destination { sender_id; _ } = Messages.private_ids [sender_id]
+
+  let destination { recipients; _ } =
+    match recipients with
+    | Stream (stream_id, topic) -> Messages.stream_id stream_id topic
+    | Private ids -> Messages.private_ids ids
+
+  (* Message filters *)
+
+  let is_trusted ?(trusted_ids = []) ?(trusted_emails = []) { sender_id; sender_email; _ } =
+    List.mem sender_id trusted_ids || List.mem sender_email trusted_emails
+
+  let has_flag flag m = List.mem flag m.flags
+
+  let is_own_message config m = m.sender_email = config.Config.email
+
+  let is_privmsg { recipients; _ } =
+    (* A message is a privmsg if it has exactly two recipients: us and the
+       sender. *)
```

Comment:
Actually you can have private messages with more than than two recipients.

Reply:
In Zulip, yes, but in this PR `ozulip` doesn't treat them as "privmsg" (see the documentation for `is_privmsg`; a "privmsg" is a private message *with a single other user*.

This is used in `interact` to determine whether to add a mention when replying (it is not needed in a conversation with a single user since the target of the message is necessarily the other user, but in a stream or in a group conversation you need to disambiguate who you are replying to.

If you have a better name than `privmsg` I'd be happy to change the name.